src/transform/src/predicate_pushdown.rs

```diff
@@ -459,95 +462,126 @@ impl PredicatePushdown {
+                    // class and collect (position within the equivalence class,
+                    // input the expression belongs to, localized version of the
+                    // expression).
+                    let mut single_input_exprs = equivalence
+                        .iter()
+                        .enumerate()
+                        .filter_map(|(pos, e)| {
+                            let mut inputs = input_mapper.lookup_inputs(e);
+                            if let Some(input) = inputs.next() {
+                                if inputs.next().is_none() {
+                                    return Some((
+                                        pos,
+                                        input,
+                                        input_mapper.map_expr_to_local(e.clone()),
+                                    ));
+                                }
+                            }
+                            None
+                        })
+                        .collect::<Vec<_>>();
+
+                    if let Some(literal_pos) = equivalence.iter().position(|expr| expr.is_literal())
+                    {
+                        // Case 2: There is one literal in the equivalence class
+                        let literal = equivalence[literal_pos].clone();
+                        let gen_literal_equality_pred = |expr| {
+                            if literal.is_literal_null() {
+                                MirScalarExpr::CallUnary {
+                                    func: expr::UnaryFunc::IsNull,
+                                    expr: Box::new(expr),
+                                }
+                            } else {
+                                MirScalarExpr::CallBinary {
+                                    func: expr::BinaryFunc::Eq,
+                                    expr1: Box::new(expr),
+                                    expr2: Box::new(literal.clone()),
+                                }
+                            }
+                        };
+                        // For every single-input expression `expr`, we can push
+                        // down `expr = literal` and remove `expr` from the
+                        // equivalence class.
+                        for (expr_pos, input, expr) in single_input_exprs.drain(..).rev() {
+                            push_downs[input].push(gen_literal_equality_pred(expr));
+                            equivalence.remove(expr_pos);
+                        }
+                        // If any multi-input expression `expr` exists in the
+                        // same equivalence class as the literal, the predicate
+                        // `expr = literal` should be
+                        // lifted above the join.
+                        for multi_input_expr in equivalence.drain(..) {
+                            lifts.push(gen_literal_equality_pred(multi_input_expr));
+                        }
```

Comment:
Why should it be lifted above the join?

Reply:
Suppose you have a query:
```
select * from
from foo, bar
where foo.a * bar.a = 5
```
The `foo.a * bar.a = 5` can either be put into a join (like it does on main currently):
```
               Optimized Plan               
--------------------------------------------
 %0 =                                      +
 | Get materialize.public.foo (u7)         +
 | ArrangeBy ()                            +
                                           +
 %1 =                                      +
 | Get materialize.public.bar (u3)         +
                                           +
 %2 =                                      +
 | Join %0 %1 (= 5 (#0 * #3))              +
 | | implementation = Differential %1 %0.()+
 | | demand = (#0..#4)                     +
 | Filter !((isnull(#0) || isnull(#3)))    +
```
Or it can live in a Filter (which is what happens on the PR)
```
               Optimized Plan               
--------------------------------------------
 %0 =                                      +
 | Get materialize.public.foo (u7)         +
 | ArrangeBy ()                            +
                                           +
 %1 =                                      +
 | Get materialize.public.bar (u3)         +
                                           +
 %2 =                                      +
 | Join %0 %1                              +
 | | implementation = Differential %1 %0.()+
 | | demand = (#0..#4)                     +
 | Filter (5 = (#0 * #3))                  +
```

I want it to be consistent whether you'd find `foo.a * bar.a = 5` in a Join or Filter after predicate pushdown. Since this PR is about making sure that stuff like `bar.a = 5` is always moved to a Filter and does not remain a join condition, I figured it would be more intuitive for `foo.a * bar.a = 5` to always be moved to a Filter.

Reply:
I'm super confused then. The logic looks like it says "if there is a literal, hoist all expressions". What happens if `equivalences` has multiple entries?

Reply:
More specifically, I guess, what happens in
```
select * from
from foo, bar
where foo.a * bar.a = 5 
  and foo.a * bar.a = foo.b * bar.b
```

Reply:
That is a good point. It turns out that on the PR branch:
```
explain 
select foo.a, bar.a 
from foo, bar 
where foo.a * bar.a = 5 
  and foo.a * bar.a = foo.b * bar.b;
                     Optimized Plan                      
---------------------------------------------------------
 %0 =                                                   +
 | Get materialize.public.foo (u7)                      +
 | ArrangeBy ()                                         +
                                                        +
 %1 =                                                   +
 | Get materialize.public.bar (u3)                      +
                                                        +
 %2 =                                                   +
 | Join %0 %1 (= (#0 * #3) (#1 * #4))                   +
 | | implementation = Differential %1 %0.()             +
 | | demand = (#0, #3)                                  +
 | Filter !((isnull(#0) || isnull(#3))), (5 = (#0 * #3))+
 | Project (#0, #3)                                     +
```

whereas
```
explain 
select foo.a, bar.a 
from foo, bar, (select 5 as a) baz 
where foo.a * bar.a = baz.a 
      and foo.a * bar.a = foo.b * bar.b;
               Optimized Plan               
--------------------------------------------
 %0 =                                      +
 | Get materialize.public.foo (u7)         +
 | ArrangeBy ()                            +
                                           +
 %1 =                                      +
 | Get materialize.public.bar (u3)         +
                                           +
 %2 =                                      +
 | Join %0 %1                              +
 | | implementation = Differential %1 %0.()+
 | | demand = (#0, #1, #3, #4)             +
 | Filter (5 = (#0 * #3)), (5 = (#1 * #4)) +
 | Project (#0, #3)                        +
```

I guess I'll make `foo.a * bar.a = 5` consistently live in the `Join` then. Code that work on joins will have to remain aware that the expressions in equivalences can refer to zero inputs.

Reply:
Do you know if leaving that in is likely to return the crash that @philip-stoev observes on `main`? If it results from this sort of issue, I'm happy to try and track down the actual point of failure and fix the offending code.

Reply:
Not to worry. The crash observed by Philip is due to the brokenness of the original predicate pushdown code.
https://github.com/MaterializeInc/materialize/blob/d6e1b2e9621b42e3f9b8ae849afeb3843ba08977/src/transform/src/predicate_pushdown.rs#L266

As you can see here, the original predicate pushdown code mistakenly thinks that the support of an `ScalarExpr` contains the inputs that the `ScalarExpr` references. Thus, it is easy to crash if the `ScalarExpr` references say, column `#12` when there are only 3 inputs. The only reason we don't crash more often is that normally, the pushdown would've happened here:

https://github.com/MaterializeInc/materialize/blob/d6e1b2e9621b42e3f9b8ae849afeb3843ba08977/src/transform/src/predicate_pushdown.rs#L207